Update sketch list styling

LakshSingla opened the pull request:
I have verified that this pull request:

* [x] has no linting errors (`npm run lint`)
* [ ] is from a uniquely-named feature branch and has been rebased on top of the latest master. (If I was asked to make more changes, I have made sure to rebase onto master then too)
* [x] is descriptively named and links to an issue number, i.e. `Fixes #390 `

@catarak I have created a layout for the dropdown menu item along with the logic for displaying the menu item from the Zeplin specs. It would be great if you could confirm the layout and styling (for white theme) :smile: . After that, I will add functionality to individual buttons.

Here are my concerns / doubts :
 - There is no highlighting for specific menu item as there is in `My Account` button. Should I add it ?
 - What colors to use in highlighting ? I have currently hardcoded hex codes as in Zeplin doc, but I guess it has to be according to theme and hence which color correspond to which theme variable. Also what about those which are not included in any theme variable. Should I create any new variable ?

![screenshot from 2019-02-02 18-24-14](https://user-images.githubusercontent.com/30999375/52164479-fd5d6c00-2717-11e9-9c9d-bb83e8db9919.png)

hellonearthis commented:
Highlighting could be done using bold text?

LakshSingla commented:
@hellonearthis I used bold text :sweat_smile:. Didn't realise it was highlighting. I will make updates in the next commit if required. 
Reference lines for others: [L103](https://github.com/LakshSingla/p5.js-web-editor/blob/54050213a13e3c7ed23b18caca0700344f862073/client/styles/components/_sketch-list.scss#L103) and [L125](https://github.com/LakshSingla/p5.js-web-editor/blob/54050213a13e3c7ed23b18caca0700344f862073/client/styles/components/_sketch-list.scss#L125)

catarak commented:
this is looking great! i'd take a closer look at the zeplin screen for some alignment/color issues:

1. there's more padding to the right of the triangle
2. the color is too light
3. it should be visible all of the time, not only on hover
4. the dropdown appears off center, too far down and too the left of the triangle icon

which highlighting do you mean? when you hover over the dropdown? i would follow the same color/highlight pattern as the navigation dropdowns.

LakshSingla commented:
@catarak I would address these details at the earliest. In the beginning I did not realize that you can view CSS styles in the Zeplin itself :sweat_smile: 
What about themifying? There is no screen(as far as I recall) corresponding to the UI in other color schemes. Would it be the same as for white ?
Also for point 2 in your comment above, by color you meant the color of the dropdown rectangle right ? :thinking: 
I would also look over integrating it with the back end and let you know my concerns.

catarak commented:
- re: themes: there's no screens for the different themes for this view, but do you best to match the colors from the other theme screens and the themes as they exist in `client/styles/abstracts/_variables.scss`
- and yes, for my point 2 above, i mean the color of the dropdown triangle on the right

LakshSingla commented:
@catarak I have made the necessary changes but I am facing some issues with `package-lock.json` post upgrading my node version. I had to delete it and reinstall `node_modules`. It may conflict with the current version of `package-lock`. I have currently commited without `package-lock` and would update it in the future commits after resolving my local issue.

meiamsome commented:
@LakshSingla Just so you're aware, it looks like you committed the deletion of the `package-lock.json` you mention, rather than just not committing the changes to that file. I highly suggest you modify the previous commit to remove that change rather than commiting your package-lock.

LakshSingla commented:
@meiamsome I have included the changed package-lock.json while updating my branch with the current master. 
@catarak, @meiamsome  I have updated the sketch list styling trying my best to incorporate the points above. I have also themifyed the dropdown menu taking references from the Nav. It would be helpful if you can provide your suggestions :). I will make the changes (if any) tomorrow and start integrating it with the backend.

catarak commented:
thanks for making the changes, it is looking better! a few things i noticed:
- the right arrow is _slightly_ off center, vertically, and needs to be moved up a little:
<img width="845" alt="screen shot 2019-02-22 at 5 41 27 pm" src="https://user-images.githubusercontent.com/6063380/53275573-8f2e1880-36c9-11e9-8221-4c77265d0ea4.png">
- when you've opened the dropdown, and you click outside of it, still on the sketch list, the whole overlay closes. it should just close the dropdown and the sketch list should still be open.
- when you hover over the last item in the list, there's a blank space at the bottom:
<img width="396" alt="screen shot 2019-02-22 at 5 49 51 pm" src="https://user-images.githubusercontent.com/6063380/53275774-5f334500-36ca-11e9-8779-b24fd62aeb46.png">
- the dark and high contrast themes have a double border between the title of the dropdown and the items, it should change the color of the dashed border:
<img width="297" alt="screen shot 2019-02-22 at 5 50 11 pm" src="https://user-images.githubusercontent.com/6063380/53275816-8be75c80-36ca-11e9-8f34-1348b76e1c3c.png">

catarak commented:
last thing, the dropdown also needs a shadow, like in the nav:
<img width="306" alt="screen shot 2019-02-22 at 5 52 35 pm" src="https://user-images.githubusercontent.com/6063380/53275850-b1746600-36ca-11e9-8469-8215b64e1fd6.png">

LakshSingla commented:
@catarak I am making the desired changes. However a few small things :
 - In the design screen (in Zeplin) there is some space at the bottom of the dropdown. (I will remove this space in my next commit. Even still, please take a look at the screens to see if that space is required back).
 - Due to the design of the SVG I will have to manually adjust a few pixels here and there to center it. I am unable to find some way that would automatically position in the center. (The div containing it is in the center though)
 - For this following point that you have raised, I am not facing any such issue. Can you please tell the operating system and browser that you are using :thinking: 
I am using Ubuntu 18.04 and Firefox Quantum( as the test browser). I will do some testing on Chrome today as well. 
>when you've opened the dropdown, and you click outside of it, still on the sketch list, the whole overlay closes. it should just close the dropdown and the sketch list should still be open.

![vid](https://user-images.githubusercontent.com/30999375/53283859-3feff280-3772-11e9-9a72-5e8c3d1a755f.gif)

LakshSingla commented:
@catarak Just to be sure, anything except `Delete` is not pre-written right ??

 - For `Rename`, `SET_PROJECT_NAME` action should suffice right ? I am confused as it only makes change in the client's side. It is not making any server side calls.

LakshSingla commented:
@catarak **Update** on overlay closing issue when we click outside :- In Google Chrome too, it is working as intended, i.e. shows the behavior as shown in the GIF commented above. (*Chrome Version 71.0.3578.98*).

LakshSingla commented:
@catarak Should I go ahead and start creating a modal for renaming of the project name (only that functionality is left in the modal) or is there another UI you have in mind. Also is the issue that `SketchList` overlay is automatically closing when clicking outside the dropdown but still on `SketchList` still recurring in your browser ? :thinking:

catarak commented:
@LakshSingla when updating a sketch name from this overlay, it should update server side. the action you need might not exist!

also, when editing a sketch name, i would follow the same UI/UX as when editing a project name when it's open—it should create an inline `<input>` element, that when you hit enter or click outside of it, it updates it.

catarak commented:
i think what's happening is that when you open the dropdown, and you click outside, it's opening a specific sketch, so it appears that it's closing the overlay, but really it's opening a sketch. kind of a confusing UX bug. i think a solution is changing how you open a sketch from the list—would need to do a little bit a research to figure out the right UX here!

LakshSingla commented:
> also, i think what's happening is that when you open the dropdown, and you click outside, it's opening a specific sketch, so it appears that it's closing the overlay, but really it's opening a sketch. kind of a confusing UX bug. i think a solution is changing how you open a sketch from the list—would need to do a little bit a research to figure out the right UX here!

Ohh I haven't thought of that case. My bad :hushed:. In most of the desktop environments, if a dropdown is open (say for example on right clicking the desktop), then any click made to the desktop is not processed, but the dropdown closes. That approach is intuitive, IMO.
So for opening a sketch of the sketchlist : 
 - *If any dropdown is not open*: Open the sketch 
 - *If any dropdown is open*: Close the dropdown 

Please suggest if you need any other type of UX :smile: . (I am not sure how such cases are handled in Macs)

I will make the said changes. Also by the `action` did you mean a React action, or server side endpoint ? (I will dig into it a little further and let you know how I will proceed)

LakshSingla commented:
@catarak I have updated the UX accordingly. Could you please check it and lemme know what you think of it :). Also, hopefully the bug that you stated is also fixed.

LakshSingla commented:
@catarak I have created the UI/UX for renaming of the project. I will write reducers and actions later. Please play around and lemme know your opinions, and suggestions (if any) I need to implement.

catarak commented:
this is looking so great! i think it's a little awkward when you click on the "share" dropdown, it appears over the sketch list, and then when you click its "X", both overlays close. i'm not sure what the right UX here though, and i think some research is necessary.

catarak commented:
and i think you'll need to use the `saveProject` action—i recently updated it so you can pass in a project rather than just pulling from the redux store.

LakshSingla commented:
@catarak According to me, `saveProject` action only works on the current project, with renaming, we need the functionality to change the name of any project. 

I am making some changes to the `updateProject` in the `projects.controller.js`. Currently, the controller only works if we pass some files in the request body, which would not be possible if we wish to just update the name of the project. More precisely, I am changing this:
`if (updatedProject.files.length !== req.body.files.length)`
to this:
`if (req.body.files && updatedProject.files.length !== req.body.files.length)`
and adding an action for renaming the project.

Lemme know if I should do this by some other way or not.
In the meantime, I will investigate the Modal Issue.

LakshSingla commented:
@catarak This feature is almost *complete* from my side :tada: :tada: 
Following are somethings that I want to mention: 
 - The module thing that you mentioned is not related to this pull, it seems to be a general behavior of closing all overlays if one seems to be present on top of other, and I feel should be addressed in a separate issue. If you also think so, lemme know, I will open another issue and start working on it. ( Otherwise, I would include that in the same issue)
 - Please take a look at my last commit and the changes I made, especially to the `server` file, and the new action I created. If you are unsatisfied with the method I employed, lemme know any other you want me to follow. The `saveProject` action, I think requires the whole project to change just the name, which I feel would be unnecessary. 
Your thoughts on this ..?

LakshSingla commented:
@catarak Here are some further issues which can be spin-off from this PR (apart from implementing the search bar, and sorting of sketches):
 - Add the rename functionality to the `File` header (as are the other options present).
 - Investigate the behavior of modals when one overlays the other. (i.e. closing of both when only one is closed)

LakshSingla commented:
@catarak I have brought my branch upto date with the current master. :smile:

LakshSingla commented:
@catarak Any further updates? Did you review the changes I made? :sweat_smile:

catarak commented:
@LakshSingla i will take a look at this today! sorry about the delay 😄

catarak commented:
just got a chance to look at this! a few things i have noticed:
- when you are looking at a list of sketches that you don't own, you have the file options "Rename" and "Delete". you should not be able to do "Download", "Duplicate", or "Share" them instead.
- sometimes, but not always, when looking at a list of sketches you do own, the only options are "Rename" and "Delete", when it should have all of the options. 
- this is a mistake in the design, but instead of the dropdown saying "File Options" at the top it should say "Sketch Actions"

still haven't had a chance to think about what should happen when clicking "share", but as it is right now it's awkward.

jonathan-s commented on client/modules/IDE/pages/IDEView.jsx:
```diff
@@ -427,6 +427,7 @@ class IDEView extends React.Component {
             <Feedback previousPath={this.props.ide.previousPath} />
           </Overlay>
         }
+        aa
```
It looks like this was added by mistake?

catarak replied:
seems like it!

LakshSingla commented:
@catarak No worries, all in good time!!. I currently have midterms going on. I will look into the changes by this weekend.

 - Regarding first and second realised what I did wrong. This might take a while to sort out. Just to be clear, I should show the users Download, Duplicate and Delete on their non owned sketches. Also, those sketches owned by the users, they should be able to see all options. (It may require some major changes I guess because one has to add support for those sketches which he is not currently viewing). 

 - I agree with the share part. I will look into the modal behaviour too.

catarak commented:
- Users should see Download, Duplicate and _Share_ on their non-owned sketches

LakshSingla commented:
@catarak So the major flaw in what I had coded was this: I was using the predefined functions for sharing, deleting etc, not realising that they work on the _current_ sketch. Hence I am currently changing my code and refactoring some components, so that they work with arbitrary ids. I apologize for any delay in the changes, there was a bunch of academic workload. I will try to complete the required functionality by tonight. _Sharing_ may require some work because the current system works with the current sketch only.

LakshSingla commented:
>     * have the file options

Regarding when I am not viewing the sketches which I own, the `SketchList.jsx` still enumerates the sketches _which I own_, not those of the person whose you are viewing, hence how does one view the list of sketches one doesnot own. Can you please clarify @catarak ??

Also, I have refactored `cloneProject` a bit. It now accepts an `id` which is the project id which you want to clone. If not provided, it defaults to the current state's id, otherwise, it queries the server for files, and then uses that to clone the project. I am pushing it in my next update. I will work on the `Share` part next.

LakshSingla commented:
@catarak My PR is  ready for another review. I have made the necessary changes. Please can you take a look at it and tell whether they are sufficient :)

jonathan-s commented on client/modules/IDE/components/SketchList.jsx:
```diff
@@ -9,13 +9,31 @@ import { bindActionCreators } from 'redux';
+import * as IdeActions from '../actions/ide';
 
-const trashCan = require('../../../images/trash-can.svg');
+// const trashCan = require('../../../images/trash-can.svg');
```
If the trashCan constant is no longer used. I'd suggest to remove this line of code entirely.

LakshSingla commented:
@jonathan-s Thanks a lot for the review, I appreciate it. I missed these lines while cleaning up my code. I will take another look at my code, to check if I have missed some other things like these. :smile:

catarak commented:
this is looking good. you've left in a few commented out changes, and it would be great if you took those out.

LakshSingla commented:
@catarak I have tried to remove all the commented out changes. Please let me know if there is any specific line which you want me to clean, or which I might have missed :smile:

jonathan-s commented on client/styles/components/_sketch-list.scss:
```diff
@@ -50,28 +59,125 @@
+        fill: getThemifyVariable('nav-hover-color');
+      }
+    }
+  }*/
```
Perhaps this commented code can be removed?

LakshSingla replied:
Thanks :)

jonathan-s commented on client/styles/components/_sketch-list.scss:
```diff
@@ -50,28 +59,125 @@
+  @include themify(){
+    border-bottom: 1px dashed map-get($theme-map, 'inactive-text-color');
+  }
+  // color: #ffffff; 
```
This code here is also commented.

LakshSingla replied:
Thanks again :)

LakshSingla commented:
Any further updates, or corrections which I may have missed ?

catarak commented:
hi @LakshSingla! sorry for the delay on this. there's been a ton of activity in this repository and i'm trying to get through it all.

LakshSingla commented:
@catarak No problem at all! I felt that something more had to be done on my part, which I forgot to address.

LakshSingla commented:
@catarak Did you get a chance to review this ?

catarak commented:
sorry for the slowness on this, hang tight!

LakshSingla commented:
@catarak Any updates still? It has been more than 1.5 months since my last commit was reviewed :sweat_smile: .

catarak commented:
i think this is ready to merge! i didn't know what to do about the share modal flow (i still don't think it's great to layer it on top of the list modal), so i decided to just comment out that option for now, for the sake of getting this merged in. thanks for your work on this @LakshSingla!

LakshSingla commented:
😊
